packages/lit-html/src/directives/style-map.ts

```diff
@@ -0,0 +1,139 @@
+/**
+ * @license
+ * Copyright (c) 2018 The Polymer Project Authors. All rights reserved.
+ * This code may only be used under the BSD style license found at
+ * http://polymer.github.io/LICENSE.txt
+ * The complete set of authors may be found at
+ * http://polymer.github.io/AUTHORS.txt
+ * The complete set of contributors may be found at
+ * http://polymer.github.io/CONTRIBUTORS.txt
+ * Code distributed by Google as part of the polymer project is also
+ * subject to an additional IP rights grant found at
+ * http://polymer.github.io/PATENTS.txt
+ */
+
+import {
+  AttributePart,
+  directive,
+  Directive,
+  noChange,
+  PartInfo,
+  ATTRIBUTE_PART,
+} from '../lit-html.js';
+
+const VENDOR_PREFIX = /^(webkit|moz|ms|o)[A-Z]/;
+
+/**
+ * A key-value set of CSS properties and values.
+ *
+ * The key should be either a valid CSS property name string, like
+ * `'background-color'`, or a valid JavaScript camel case property name
+ * for CSSStyleDeclaration like `backgroundColor`.
+ */
+export interface StyleInfo {
+  readonly [name: string]: string;
+}
+
+class StyleMap extends Directive {
+  previousStyleProperties?: Set<string>;
+
+  constructor(part: PartInfo) {
+    super();
+    if (
+      part.type !== ATTRIBUTE_PART ||
+      part.name !== 'style' ||
+      (part.strings !== undefined && part.strings.length > 2)
+    ) {
+      throw new Error(
+        'The `styleMap` directive must be used in the `style` attribute ' +
+          'and must be the only part in the attribute.'
+      );
+    }
+  }
+
+  render(styleInfo: StyleInfo) {
+    return Object.entries(styleInfo)
+      .reduce((style, [prop, value]) => {
+        if (value === null) {
+          return style;
+        }
+        // Convert property names from camel-case to dash-case, i.e.:
+        //  `backgroundColor` -> `background-color`
+        // Vendor-prefixed names need an extra `-` appended to front:
+        //  `webkitAppearance` -> `-webkit-appearance`
+        // Exception is any property name containing a dash, including
+        // custom properties; we assume these are already dash-cased i.e.:
+        //  `--my-button-color` --> `--my-button-color`
+        // Note that prefixed
+        if (prop.indexOf('-') === -1) {
+          if (VENDOR_PREFIX.test(prop)) {
+            prop = '-' + prop;
+          }
+          prop = prop.replace(/([A-Z])/g, '-$1').toLowerCase();
+          return style + `${prop}: ${value}; `;
+        }
+        return style + `${prop}:${value}; `;
+      }, '')
+      .trim();
```

Comment:
I think I added the `trim()` in when writing tests for SSR to avoid the string-based render differing from the serialization the browser does to the attribute when modifying it via the OM (this was also the reason for the space between properties). I sort of forget if there was a case where this would register as an attribute mutation, or if it was to just avoid encoding the exact serialization rules the string render did in the test. We could probably remove it for now and add it back in if necessary when we port the SSR stuff over.